design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,243 @@
+
+## Goals
+
+* Enable schema documentation and validation for resource classes.
+* Continue to support configuring a default resource class to be set upon resource claims that do not specify a class.
+* Assume that the universe of possible default resource class kinds for a specific claim kind will grow arbitrarily, so a default class controller must be able to be made aware of new resource class kinds. This goal will become specifically important as we begin to [separate providers](https://github.com/crossplaneio/crossplane/issues/531) from the core Crossplane project in the form of infrastructure stacks. 
+
+## Proposal
+
+Unfortunately, strongly typed resource classes make it much more difficult to identify default resource classes. Currently, there is one default class controller per claim kind that watches for creation of their specified claim kind, then lists objects of `kind: ResourceClass` that are labeled as default for the claim kind. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller can no longer list a single `ResourceClass` kind.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new set of CRD's can be implemented to specify when an instance of a strongly typed resource class kind should serve as default for a given claim kind. These default policies should be able to be defined at both the cluster and namespace level, so there must exist two separate CRD's for each claim kind. For example, the `MySQLInstance` claim kind will be accompanied by the `MySQLInstancePolicy` (namespace-scoped) and `MySQLInstanceClusterPolicy` (cluster-scoped) kinds. Importantly, when single instances of both the `MySQLInstancePolicy` and `MySQLInstanceClusterPolicy` exist, the `MySQLInstancePolicy` would take precedence because it is at the more granular (namespace) level.
+
+## Workflow
+
+The workflow for a namespace-scoped policy (`kind: MySQLInstancePolicy`) would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRD's and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+spec:
```

Comment:
`spec` feels a little strange here, since it's not specifying the configuration of the resource class. I'd suggest either something like `specTemplate`, or the pattern used by `Deployment` and friends, e.g.:

```yaml
apiVersion: database.aws.crossplane.io/v1alpha1
kind: RDSInstanceClass
metadata:
  name: standard-mysql
  namespace: crossplane-system
template:
  spec:
    class: db.t2.small
    # ...
```

Reply:
Agreed, I am in favor of the latter.

Reply:
Since this situation is somewhat different than that of a `Deployment`, we have decided on `specTemplate`.